nemo/core/neural_modules.py

```diff
@@ -37,47 +38,137 @@ class WeightShareTransform(Enum):
 
         # Set module properties from factory else use defaults
-        self._placement = factory.placement if factory is not None else DeviceType.GPU
-        self._opt_level = factory.optim_level if factory is not None else Optimization.mxprO0
+        self._placement = self._factory.placement
+        # If one needs to change that should override it manually.
 
-        # Update module properties using overrides if overrides exist
-        if placement is not None:
-            self._placement = placement
+        # Optimization level.
+        self._opt_level = self._factory.optim_level
 
-        self._factory = factory
+        # Get object UUID.
         self._uuid = str(uuid.uuid4())
 
-        # if kwargs:
-        #    nemo.logging.warning(
-        #        "When constructing {}. The base "
-        #        "NeuralModule class received the following unused "
-        #        "arguments:".format(self.__class__.__name__))
-        #    nemo.logging.warning("{}".format(kwargs.keys()))
+        # Retrieve dictionary of parameters (keys, values) passed to init.
+        self._init_params = self.extract_init_params()
+
+        # Pint the types of the values.
+        # for key, value in self._init_params.items():
+        #    print("{}: {} ({})".format(key, value, type(value)))
+
+        # Validate the parameters.
+        # self.validate_params(self._init_params)
+
+    def extract_init_params(self):
+        """
+            Retrieves the dictionary of of parameters (keys, values) passed to constructor of a class derived
+            (also indirectly) from the Neural Module class.
+
+            Returns:
+                Dictionar containing parameters passed to init().
+        """
+        # print("Hello from extract_init_params")
+        # print(type(self))
+        # print(getfullargspec(type(self).__init__))
+
+        # Get names of arguments of the original module init method.
+        init_keys = getfullargspec(type(self).__init__).args
+
+        # Remove self.
+        if "self" in init_keys:
+            init_keys.remove("self")
+
+        # Create list of params.
+        init_params = {}.fromkeys(init_keys)
+
+        # Retrieve values of those params from the call list.
+        for frame in stack()[1:]:
+            localvars = getargvalues(frame[0]).locals
+            # print("localvars: ", localvars)
+            for key in init_keys:
+                # Found the variable!
+                if key in localvars.keys():
+                    # Save the value.
+                    init_params[key] = localvars[key]
+
+        # Return parameters.
+        return init_params
+
+    @property
+    def init_params(self) -> Optional[Dict]:
+        """
+            Property returning parameters used to instantiate the module.
+
+            Returns:
+                Dictionary containing parameters used to instantiate the module.
+        """
+        return self._init_params
+
+    def validate_params(self, params):
```

Comment:
should be private. But I missed it's use - shouldn't it be called inside the constructor?

Reply:
actually, I think it's better to take this out of NeuralModule class.
So two options - you chose: (1) make private (and static?) and (2) put into some utils

Reply:
Ok, thanks, the method will be used in export to config file, will decide about that later.